### PR TITLE
Use math and cmath instead of numpy on scalars

### PR DESCRIPTION
**Pull request by levbishop**
There's no reason to use numpy functions instead of math/cmath for scalars and the latter are around 10x faster due to avoiding the numpy dispatch overhead. This PR replaces a bunch of instances. I don't expect any sizeable performance impact as most are not in any core loop but there may be some minor benefit and I'd rather have more examples of best-practice in the codebase. One slight difference between math and numpy is that floor/ceil do already return integer so no need to wrap int(np.floor(x))

### CODE REVIEW

File: qiskit/circuit/gate.py [hunk 1]
```diff
@@ -42,7 +42,6 @@ def __init__(
             params: A list of parameters.
             label: An optional label for the gate.
         """
-        self.definition = None
```
Comment (jakelishman) [inline qiskit/circuit/gate.py]: This seems unrelated (and potentially a dangerous change in the subclass relationship? - I _hope_ it isn't, but I know better than to trust that kind of thing in Qiskit.).

File: qiskit/quantum_info/states/utils.py [hunk 2]
```diff
@@ -102,17 +103,18 @@ def shannon_entropy(pvec: list | np.ndarray, base: int = 2) -> float:
...
         def logfn(x):
-            return -x * np.log(x)
+            return -x * math.log(x)
 
     else:
+        log_base = math.log(base)
 
         def logfn(x):
-            return -x * np.log(x) / np.log(base)
+            return -x * math.log(x) / log_base
```
Comment (jakelishman) [inline qiskit/quantum_info/states/utils.py:117]: This is a weird codeblock.  It could fairly easily have been vectorised (`scipy.special.xlogy`), and the `log_base` multiplier could have been brought outside the loop to avoid needing to make the function virtual.

Anyway, irrelevant to this PR.

File: qiskit/synthesis/discrete_basis/commutator_decompose.py [hunk 3]
```diff
@@ -92,8 +92,8 @@ def _solve_decomposition_angle(matrix: np.ndarray) -> float:
     lhs = math.sin(angle / 2)
 
     def objective(phi):
-        sin_sq = np.sin(phi / 2) ** 2
-        return 2 * sin_sq * np.sqrt(1 - sin_sq**2) - lhs
+        sin_sq = math.sin(phi[0] / 2) ** 2
+        return 2 * sin_sq * math.sqrt(1 - sin_sq**2) - lhs
```
Comment (jakelishman) [inline qiskit/synthesis/discrete_basis/commutator_decompose.py]: This looks like a behavioural change - how sure are we that `phi` is a single item?  Perhaps `phi.item()` is better if we're sure, because it'll fail noisily if it's not.

File: qiskit/synthesis/unitary/aqc/__init__.py [hunk 4]
```diff
@@ -108,7 +108,7 @@
     unitary = ...
 
     # Define a number of qubits for the algorithm, at least 3 qubits
-    num_qubits = int(round(np.log2(unitary.shape[0])))
+    num_qubits = int(round(math.log2(unitary.shape[0])))
```
Comment (jakelishman) [inline qiskit/synthesis/unitary/aqc/__init__.py]: `int(round(...))` seems very wrong, but that's extra this PR

**Review by jakelishman — COMMENTED** [review]
Thanks for this - aside from a comment or two this all looks good and it's good to get rid of the Numpy overhead.

File: qiskit/circuit/gate.py (hunk 1 above)
Comment (levbishop) [reply qiskit/circuit/gate.py]: Oops. I didn't mean to include this here, but I think it's safe since `Instruction.__init__` sets `self._definition = None` and it saves some measurable time on this by avoiding a call to `@definition.setter`. I'll shove to a separate PR

File: qiskit/quantum_info/states/utils.py (hunk 2 above)
Comment (levbishop) [reply qiskit/quantum_info/states/utils.py:117]: agree its a weird block

File: qiskit/circuit/gate.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/circuit/gate.py]: The PR that introduced `ControlledGate` deliberately included this line, which is what makes me particularly nervous about it, but if it passes the tests, I think it ought to be fine.

File: qiskit/synthesis/unitary/aqc/__init__.py (hunk 4 above)
Comment (jakelishman) [reply qiskit/synthesis/unitary/aqc/__init__.py]: "extra _to_ this PR", I mean - no need to change it.

File: qiskit/circuit/gate.py (hunk 1 above)
Comment (levbishop) [reply qiskit/circuit/gate.py]: Dropped this change in af57ee4cb4c421d922c16b22cf69a1bc85efdf98

File: qiskit/synthesis/discrete_basis/commutator_decompose.py (hunk 3 above)
Comment (levbishop) [reply qiskit/synthesis/discrete_basis/commutator_decompose.py]: we are sure its a single item as its only called from `fsolve(objective, angle)` where `angle` is scalar, but I changed to `phi.item()` for ease of future understanding in af57ee4cb4c421d922c16b22cf69a1bc85efdf98

File: qiskit/synthesis/unitary/aqc/__init__.py (hunk 4 above)
Comment (levbishop) [reply qiskit/synthesis/unitary/aqc/__init__.py:111]: dropped the int since its superfluous in af57ee4cb4c421d922c16b22cf69a1bc85efdf98

Comment (levbishop) [reply qiskit/synthesis/unitary/aqc/__init__.py]: Fixed it anyway  in 9588d6c9e

File: qiskit/synthesis/discrete_basis/commutator_decompose.py (hunk 3 above)
Comment (jakelishman) [reply qiskit/synthesis/discrete_basis/commutator_decompose.py]: I'm kind of surprised that `fsolve` allows the objective function to return a differently-shaped array to what it passes in, but as long as it works, I'm good with it.

File: qiskit/synthesis/two_qubit/xx_decompose/circuits.py [hunk 5]
```diff
@@ -291,8 +292,8 @@ def canonical_xx_circuit(target, strength_sequence, basis_embodiments):
         circuit.compose(basis_embodiments[strength_sequence[0]], inplace=True)
         circuit.compose(source_reflection.inverse(), inplace=True)
         circuit.compose(source_shift, inplace=True)
-        circuit.global_phase += -np.log(shift_phase_shift).imag
-        circuit.global_phase += -np.log(reflection_phase_shift).imag
+        circuit.global_phase -= cmath.phase(shift_phase_shift)
+        circuit.global_phase -= cmath.phase(reflection_phase_shift)
```
Comment (jakelishman) [inline qiskit/synthesis/two_qubit/xx_decompose/circuits.py:296]: We're strongly assuming that the things marked `phase_shift` have unit magnitude, I assume?  Feels like this change could bring about a change in the exact numerics, though if we were particularly worried about that, feels like we ought to calculate with the phase shifts stored as angles directly.

I'm fine if this passes the test suite, but might cause some minor merge conflicts with #12105.

**Review by jakelishman — APPROVED** [review]
Thanks for doing this, and for the follow-up changes.  Possibly let's just leave the test suite be?  I imagine there's jillions more occurrences in there, and I'm not sure I've got the strength to read all of those haha

File: qiskit/synthesis/two_qubit/xx_decompose/circuits.py (hunk 5 above)
Comment (levbishop) [reply qiskit/synthesis/two_qubit/xx_decompose/circuits.py:296]: even if not unit magnitude `cmath.phase(scalar)` and `np.log(scalar).imag` should be the same

Comment (levbishop) [reply qiskit/synthesis/two_qubit/xx_decompose/circuits.py:296]:
```ipython
In [26]: q = 100.+22.2j

In [27]: cmath.phase(q) - np.log(q).imag
Out[27]: 0.0
```

Comment (jakelishman) [reply qiskit/synthesis/two_qubit/xx_decompose/circuits.py:296]: oh yeah true sorry, I was just thinking about the logarithm in general, which is obviously nonsense

**Comment by levbishop** [comment]
> Thanks for doing this, and for the follow-up changes. Possibly let's just leave the test suite be? I imagine there's jillions more occurrences in there, and I'm not sure I've got the strength to read all of those haha

Theres probably ~200 in test/ (and likely a bunch more in qiskit/ that I couldn't immediately spot that they act on scalar).

Future patches accepted...
